src/tss2-esys/esys_tcti_default.c

```diff
@@ -1,56 +1,214 @@
+    if (*tcti == NULL) {
+        LOG_ERROR("Memory allocation for tcti failed: %s", strerror(errno));
+        return TSS2_ESYS_RC_MEMORY;
+    }
+
+    r = init(*tcti, &size, conf);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_WARNING("TCTI init for function %p failed with %" PRIx32, init, r);
+        free(*tcti);
+        *tcti=NULL;
+        return r;
+    }
+
+    LOG_DEBUG("Initialized TCTI for config: %s", conf);
+
+    return TSS2_RC_SUCCESS;
+}
+
+static TSS2_RC
+tcti_from_info(TSS2_TCTI_INFO_FUNC infof,
+               const char* conf,
+               TSS2_TCTI_CONTEXT **tcti)
+{
+    TSS2_RC r;
+    LOG_TRACE("Attemting to load TCTI info");
+
+    const TSS2_TCTI_INFO* info = infof();
+    if (info == NULL) {
+        LOG_ERROR("TCTI info function failed");
+        return TSS2_ESYS_RC_GENERAL_FAILURE;
+    }
+    LOG_TRACE("Loaded TCTI info named: %s", info->name);
+    LOG_TRACE("TCTI description: %s", info->description);
+    LOG_TRACE("TCTI config_help: %s", info->config_help);
+
+    r = tcti_from_init(info->init, conf, tcti);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_WARNING("Could not initialize TCTI named: %s", info->name);
+        return r;
+    }
+
+    LOG_DEBUG("Initialized TCTI named: %s", info->name);
+
+    return TSS2_RC_SUCCESS;
+}
+
+#ifndef NO_DL
+static TSS2_RC
+tcti_from_file(const char *file,
+               const char* conf,
+               TSS2_TCTI_CONTEXT **tcti)
+{
+    TSS2_RC r;
+    void *handle;
+    TSS2_TCTI_INFO_FUNC infof;
+
+    LOG_TRACE("Attemting to load TCTI file: %s", file);
+
+    handle = dlopen(file, RTLD_NOW);
+    if (handle == NULL) {
+        LOG_WARNING("Could not load TCTI file: %s", file);
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    infof = (TSS2_TCTI_INFO_FUNC) dlsym(handle, TSS2_TCTI_INFO_SYMBOL);
+    if (infof == NULL) {
+        LOG_ERROR("Info not found in TCTI file: %s", file);
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    r = tcti_from_info(infof, conf, tcti);
+    if (r != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Could not initialize TCTI file: %s", file);
+        dlclose(handle);
+        return r;
+    }
+
+    LOG_DEBUG("Initialized TCTI file: %s", file);
+
     return TSS2_RC_SUCCESS;
 }
+#endif /* NO_DL */
+
+TSS2_RC
+get_tcti_default(TSS2_TCTI_CONTEXT ** tcticontext)
+{
+    if (tcticontext == NULL) {
+        LOG_ERROR("tcticontext must not be NULL");
+        return TSS2_TCTI_RC_BAD_REFERENCE;
+    }
+    *tcticontext = NULL;
+#ifdef ESYS_TCTI_DEFAULT_MODULE
+
+#ifdef ESYS_TCTI_DEFAULT_CONFIG
+    const char *config = _XSTR(ESYS_TCTI_DEFAULT_CONFIG);
+#else /* ESYS_TCTI_DEFAULT_CONFIG */
+    const char *config = "";
+#endif /* ESYS_TCTI_DEFAULT_CONFIG */
+
+    LOG_DEBUG("Attempting to initialize TCTI defined during compilation: %s:%s",
+              ESYS_TCTI_DEFAULT_MODULE, config);
+    return tcti_from_file(ESYS_TCTI_DEFAULT_MODULE, config, tcticontext);
+
+#else /* ESYS_TCTI_DEFAULT_MODULE */
```

Comment:
Is this else needed. If ESYS_TCTI_DEFAULT_MODULE is defined we call tcti_from_file(ESYS_TCTI_DEFAULT_MODULE,..) and return.

Reply:
Exactly. Thus code-scanners could complain about unreachable code after return. Thus the #else...